prov/util/src/util_mr.c

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2016 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include <stdlib.h>
+#include <fi_enosys.h>
+#include <fi_util.h>
+
+
+/* deep copy: make seperate copy of mr_attr and use context for prov_mr */
+static inline util_mr_item_t * create_mr_attr_copy(
+                        const struct fi_mr_attr *in_attr, void * prov_mr)
+{
+    util_mr_item_t * item;
+    struct iovec *mr_iov;
+    int i = 0;
+
+    assert(prov_mr && in_attr);
+
+    item = malloc(sizeof(util_mr_item_t));
+    *item = *in_attr;
+    item->context = prov_mr;
+    mr_iov = malloc(sizeof(struct iovec) * in_attr->iov_count);
+
+    for(i = 0; i < in_attr->iov_count; i++)
+        mr_iov[i] = in_attr->mr_iov[i];
+
+    item->mr_iov = mr_iov;
+
+    return item;
+}
+
+static inline uint64_t get_mr_key(struct util_mr *mr_h)
+{
+    assert(mr_h->b_key != UINT64_MAX);
+    uint64_t b_key = mr_h->b_key;
+    mr_h->b_key++;
+    return b_key;
+}
+
+static inline int verify_addr(util_mr_item_t * item, uint64_t in_access,
+                                 uint64_t in_addr, ssize_t in_len)
+{
+    int i = 0;
+
+    if(!in_addr)
+        return -FI_EINVAL;
+
+    if((in_access & item->access) != in_access)
+        return -FI_EACCES;
+
+      for (i = 0; i < item->iov_count; i++) {
+        if ((uint64_t)item->mr_iov[i].iov_base <= in_addr &&
```

Comment:
Something like this may be clearer:

```c
uint64_t start = item->mr_iov[i].iov_base;
uint64_t end = start + item->mr_iov[i].iov_len;

if (start <= in_addr && end >= (in_addr + in_len))
    return 0;
```